enhancements/kube-apiserver/startup-monitor.md

```diff
@@ -0,0 +1,219 @@
+    - **source invalid data from other sources with incomplete validation**: e.g. the network mask given through some other config.openshift.io/v1 CR defined by the user is invalid.
+- **some ConfigMap or Secret can be wrong**: e.g. the audit policy is a YAML file inside a ConfigMap. If it is invalid, the kube-apiserver will refuse startup. ConfigMaps and Secrets are synced into the operand namespace from many different sources, and then copied as files onto the master node file system by the installer pod.
+- **some ConfigMap or Secret marked as optional but is actually required**: the installer will continue rolling out the pod after skipping the optional files. In HA OpenShift, this might not show up because the operator create quickly another revision when the ConfigMap or Secret show up. In SNO this race might be fatal.
+
+It is impossible to guarantee that none of these ever happens because the test permutations are just infeasible to cover completely in CI, especially as many of these are disruptive for cluster behaviour and hence very expensive to test.
+
+### Goals
+
+- recover from a badly formatted or invalid pod manifest
+- recover from bad config observer output
+- recover from bad revisioned ConfigMaps and Secrets.
+
+### Non-Goals
+
+- recover from invalid non-revisioned certificates
+- recover from missing or invalid files on the masters given by static paths outside of the `/etc/kubernetes` directory (e.g. some certs or kubeconfig the operand is consuming)
+- optimize downtime of the API beyond in fallback case to O(60s). This mechanism is for the disaster case and only survival counts. If downtime is `O(5min)` or even more, this is completely ok.
+
+## Proposal
+
+We propose to add a `<operand>-startup-monitor` static pod that watches the operand pods for readiness (see below for details) and that is created by installer as another manifest in `/etc/kubernetes` and provides by the static pod controller the same way the operand pod manifest is created today.
+
+If the operand pod does not start up in N minutes, e.g. due to:
+
+- invalid pod manifest
+- free port wait loop timeout
+- crash-looping
+- not answering on the expected port (connection refused)
+- healthy but never readyz
+
+the task for startup-monitor is to fall back:
+
+1. when detecting problems with the new revision, the startup-monitor will copy the pod-manifest of the `/etc/kubernetes/static-pods/last-known-good` link (or the previous revision if the link does not exist, or don't do anything if there is no previous revision as in bootstrapping) into `/etc/kubernetes`.
+
+It will add annotations to the old revision pod manifest:
```

Comment:
is having annotation enough? how would that be discovered by a human?

if admin make config change that that change is being reverted by non-ready API server, how do they know they did something wrong?

Reply:
The operator will publish this in degraded annotation.

Reply:
clarified.